### ISSUE TITLE
Implement spread and dynspread for aggregates, not just images

### ISSUE THREAD

**Issue by FelixLouis** [holoviz/datashader#326]
Both datashader.transfer_functions.spread and datashader.transfer_functions.dynspread have a "how" argument, set by default to "over".

The problem is that there is no explanation in the documentation (http://datashader.readthedocs.io/en/latest/api.html#datashader.transfer_functions.spread) about what values this argument can take and the effects of these values. After looking through the source code I finally found 4 possible values for this argument (https://github.com/bokeh/datashader/blob/master/datashader/composite.py):

 - "over", the default one, which seems to work so that each pixel go over the other ones with a priority set in function of their coordinates.
 - "source", which seems to work the same way (I couldn't determine the difference by testing)
 - "add", which I expected to add the values of the pixels but seems to works differently when a pixel already at the maximum value collides with another one.
 - "saturate", which I expected to take the maximum of the two pixels colliding but has the same weird behavior as "add".

**Comment by jbednar** [comment]
Sorry about that! Clearly an oversight.  The various possible operators are illustrated nicely at https://www.cairographics.org/operators/ .  We have only implemented a subset, as you list above.  Unfortunately, for RGB data there is a fixed upper bound on intensity for each channel, which makes many of the operators behave in unexpected ways.  As a result, none of them are particularly great; you should definitely try to avoid spreading that overlaps in the first place.  If you have any ideas for a better approach, adding a new operator should be trivial...

Actually, based on that link, maybe there is a problem with "over"?  It is meant to be different from "source"...

**Comment by FelixLouis** [comment]
Here is an example of what "over" can yield: If I understand correctly this is supposed to be "source" behavior (here it's a dynspread with a threshold of 1).
![over_behavior](https://cloud.githubusercontent.com/assets/25052300/25527132/9325d946-2c17-11e7-8c8c-98c595cf4470.png)

Thanks to your link I realize that what I would need is to overlap not by adding the colors but the values of the matrix just before .shade convert it to an image. For example I would like a 1px square spread to transform this:
[
[0,0,0,1],
[0,0,0,1],
[0,0,0,0],
[0,1,0,0]
]
into this:
[
[0,0,2,2],
[0,0,2,2],
[1,1,2,1],
[1,1,1,0]
]
so basically I would need to modify the canvas and not the picture. I didn't notice such function, though, is there one I could have missed? If not, is there a more efficient way to do that than the naive approach I can think about which would consist in iterating over the whole canvas with datashader.glyphs.Point?

**Comment by jbednar** [comment]
We'd be happy to have spreading at the aggregate level rather than the image level, which could indeed be useful, but have not previously implemented that.  This can be implemented as a simple transfer function operating over the output of Canvas.points(), which returns an xarray.  Such a function would need to make sure that it can handle both 2D and 3D aggregates (working on each 2D slice of the 3D aggregate, but otherwise should be simple enough to write using the same mechanism as already implemented for images (but simpler).  I'd be happy to consider a PR achieving that!

**Comment by FelixLouis** [comment]
Noted! I can't promise I'll be able to submit a valuable PR about that but if I come up with something it will be with pleasure.

**Comment by jbednar** [comment]
Implemented in #771 and released!
